src/lib/base64.c

```diff
@@ -46,35 +46,36 @@ static const unsigned char b64dec[256] = {
 void base64_encode(const void *src, size_t src_size, buffer_t *dest)
 {
 	const unsigned char *src_c = src;
-	unsigned char tmp[4];
-	size_t src_pos;
-
-	for (src_pos = 0; src_pos < src_size; ) {
-		tmp[0] = b64enc[src_c[src_pos] >> 2];
-		switch (src_size - src_pos) {
-		case 1:
-			tmp[1] = b64enc[(src_c[src_pos] & 0x03) << 4];
-			tmp[2] = '=';
-			tmp[3] = '=';
-			src_pos++;
-			break;
-		case 2:
-			tmp[1] = b64enc[((src_c[src_pos] & 0x03) << 4) |
-					(src_c[src_pos+1] >> 4)];
-			tmp[2] = b64enc[((src_c[src_pos+1] & 0x0f) << 2)];
-			tmp[3] = '=';
-			src_pos += 2;
-			break;
-		default:
-			tmp[1] = b64enc[((src_c[src_pos] & 0x03) << 4) |
-					(src_c[src_pos+1] >> 4)];
-			tmp[2] = b64enc[((src_c[src_pos+1] & 0x0f) << 2) |
-					((src_c[src_pos+2] & 0xc0) >> 6)];
-			tmp[3] = b64enc[src_c[src_pos+2] & 0x3f];
-			src_pos += 3;
-			break;
-		}
-		buffer_append(dest, tmp, 4);
+	const size_t res_size = (src_size / 3 + (src_size % 3 > 0)) * 4;
+	unsigned char *ptr = buffer_append_space_unsafe(dest, res_size);
+	
+	size_t src_pos = 0;
+	for (; src_size - src_pos > 2; src_pos += 3) {
+		*ptr++ = b64enc[src_c[src_pos] >> 2];
```

Comment:
I think it'd be worth it to avoid pointer arithmetic and use an index into the output to make it more obvious that there isn't a chance of an overflow.  I'm thinking something like:

```C
size_t src_pos = 0;
size_t dst_pos = 0;
for (; src_size - src_pos > 2; src_pos += 3, dst_pos += 4) {
        ptr[dst_pos + 0] = ...
        ptr[dst_pos + 1] = ...
        ptr[dst_pos + 2] = ...
        ptr[dst_pos + 3] = ...
}
```

And just to be on the safe side, after the loop I'd throw in an `i_assert` to make sure that `dst_pos` didn't go beyond `res_size`.  (I think `i_assert(dst_pos <= res_size)` should do.)

Reply:
Sure, indexed access shouldn't impact optimized machine code. I'll rewrite it.